service/src/main/java/bio/terra/axonserver/service/cromwellworkflow/CromwellWorkflowService.java

```diff
@@ -121,6 +151,142 @@ public CromwellApiWorkflowQueryResponse getQuery(
+   * @param workflowType Workflow language for the submitted file (i.e., WDL)
+   * @param workflowTypeVersion Version for the workflow language (draft-2 or 1.0).
+   * @param labels JSON string of labels.
+   * @param workflowDependenciesGcsUri URI pointing to the workflow dependencies: a GCS object that
+   *     is a ZIP file.
+   * @param requestedWorkflowId An ID to ascribe to this workflow. If not supplied, then a random ID
+   *     will be generated.
+   * @param token Bearer token.
+   * @return The workflow ID, and status of submission.
+   * @throws bio.terra.cromwell.client.ApiException Exception thrown by Cromwell client.
+   * @throws IOException Exception thrown during file input/output.
+   */
+  public CromwellApiWorkflowIdAndStatus submitWorkflow(
+      UUID workspaceId,
+      String workflowGcsUri,
+      String workflowUrl,
+      Boolean workflowOnHold,
+      String workflowInputs,
+      Map<String, Object> workflowOptions,
+      String workflowType,
+      String workflowTypeVersion,
+      String labels,
+      String workflowDependenciesGcsUri,
+      UUID requestedWorkflowId,
+      BearerToken token)
+      throws bio.terra.cromwell.client.ApiException, IOException {
+
+    if (workflowGcsUri == null && workflowUrl == null) {
+      throw new ApiException("workflowGcsUri or workflowUrl needs to be provided.");
+    }
+
+    // Create temp files from JSON of: workflowInputs, workflowOptions, labels, source,
+    // and dependencies.
+    // - labels and options will be modified before being sent to Cromwell.
+
+    File tempInputsFile = null;
+    if (workflowInputs != null) {
+      tempInputsFile = createSafeTempFile("workflow-label-", "-terra");
+      try (OutputStream out = new FileOutputStream(tempInputsFile)) {
+        out.write(workflowInputs.getBytes(StandardCharsets.UTF_8));
+      }
+    }
+
+    File tempOptionsFile = createSafeTempFile("workflow-options-", "-terra");
+    // Adjoin preset options for the options file.
+    // Place the project ID + compute SA into the options.
+    String projectId = wsmService.getGcpContext(workspaceId, token.getToken()).getProjectId();
+    workflowOptions.put("google_project", projectId);
+    workflowOptions.put(
+        "google_compute_service_account", samService.getPetServiceAccount(projectId, token));
+    workflowOptions.put(
+        "default_runtime_attributes", new AbstractMap.SimpleEntry("docker", "debian:stable-slim"));
+
+    ObjectMapper mapper = new ObjectMapper();
+    try (OutputStream out = new FileOutputStream(tempOptionsFile)) {
+      out.write(mapper.writeValueAsString(workflowOptions).getBytes(StandardCharsets.UTF_8));
+    }
+
+    File tempLabelsFile = createSafeTempFile("workflow-labels-", "-terra");
+    // Adjoin the workspace-id label to the workflow.
+    JSONObject jsonLabels = labels == null ? new JSONObject() : new JSONObject(labels);
+    jsonLabels.put(WORKSPACE_ID_LABEL_KEY, workspaceId);
```

Comment:
Just to confirm, this will replace a prior key with this label right?

Reply:
Yes.